hw/drivers/ipc_nrf5340/src/ipc_nrf5340.c

```diff
@@ -0,0 +1,243 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include <os/os.h>
+#include <ipc_nrf5340/ipc_nrf5340.h>
+#include <nrfx.h>
+
+/* TODO for now this allows only for 1-1 connection */
+
+#define IPC_MAX_CHANS MYNEWT_VAL(IPC_NRF5340_CHANNELS)
+#define IPC_BUF_SIZE (MYNEWT_VAL(IPC_NRF5340_BUF_SZ) + 1)
+
+struct ipc_channel {
+    ipc_nrf5340_recv_cb cb;
+    void *user_data;
+};
+
+struct ipc_shm {
+    volatile uint16_t head;
+    volatile uint16_t tail;
+    uint8_t buf[IPC_BUF_SIZE];
+};
+
+static struct ipc_channel ipcs[IPC_MAX_CHANS];
+
+#if MYNEWT_VAL(BSP_NRF5340)
+static struct ipc_shm __attribute__((section (".ipc"))) shms[IPC_MAX_CHANS];
+#elif MYNEWT_VAL(BSP_NRF5340_NET)
+static struct ipc_shm *shms = (void *)0x20000400;
+#else
+#error "Unsupported CPU"
+#endif
+
+static uint16_t
+ipc_nrf5340_shm_get_data_length(uint16_t head, uint16_t tail)
+{
+    if (head >= tail) {
+        return head - tail;
+    }
+
+    return IPC_BUF_SIZE - (tail - head);
+}
+
+static void
+ipc_nrf5340_shm_write(struct ipc_shm *shm, const void *data, uint16_t data_len)
+{
+    uint16_t head = shm->head;
+    uint16_t tail = shm->tail;
+    uint16_t len;
+
+    if (data_len == 0) {
+        return;
+    }
+
+    /* check if full */
+    assert((head + 1) != tail);
+    assert(data_len + ipc_nrf5340_shm_get_data_length(head, tail) < IPC_BUF_SIZE);
```

Comment:
If those asserts are a way to halt execution when write can't be performed. First one could be removed since it is also covered by second check.
If on the other hand they are here for development sake to assert what are conditions (and could be later removed by NDEBUG) then function will fail when Mynewt finally adopt convention that assert no code when NDEBUG is defined.